examples/eSealing/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+//! registered the file hash.
+//!
+//! Every account can only register each file hash
+//! once. This is because we want to record the time when the witness
+//! (sender_account) first had access to that file. Re-registering a file hash
+//! by the same account would not give us any additional information.
+//!
+//! But: A second account can register a file hash that was already previously
+//! registered. This second account will be added as another witness together
+//! with the timestamp when the second account called the `register_file`
+//! function. We know that the second account is most likely now also in
+//! possession of the file if it has used this eSealing service via the
+//! front-end. But this is not guaranteed since the second account could
+//! register a random file hash it found in the events by interacting with the
+//! smart contract directly (without using the front-end).
+use concordium_std::*;
+use std::ops::Deref;
+
+/// The different errors the contract can produce.
+#[derive(Serialize, Debug, PartialEq, Eq, Reject, SchemaType)]
+enum ContractError {
+    /// Failed parsing the parameter.
+    #[from(ParseError)]
+    ParseParams,
+    /// Failed logging: Log is full.
+    LogFull,
+    /// Failed logging: Log is malformed.
+    LogMalformed,
+    /// Only accounts can register a file hash.
+    OnlyAccount,
+    /// Every account can only register each file hash once.
+    AlreadyRecorded,
+}
+
+/// Mapping the logging errors to ContractError.
+impl From<LogError> for ContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// The state tracked for each file.
+#[derive(Serial, DeserialWithState, Deletable, StateClone)]
+#[concordium(state_parameter = "S")]
+struct FileState<S> {
+    /// The timestamps when this file hash was recorded.
+    timestamps:    Vec<Timestamp>,
+    /// The witnesses (sender_accounts) that recorded this file hash.
+    witnesses:     Vec<AccountAddress>,
+    /// The witnesses (sender_accounts) stored in a set. This
+    /// set enables us to check that a witness has not been recorded
+    /// without iterating through the above array which
+    /// saves execution time in the `register_file` function.
+    witnesses_set: StateSet<AccountAddress, S>,
+}
+
+/// The contract state.
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    files: StateMap<String, FileState<S>, S>,
+}
+
+impl<S: HasStateApi> State<S> {
+    /// Creates a new state with no files recorded.
+    fn new(state_builder: &mut StateBuilder<S>) -> Self {
+        State {
+            files: state_builder.new_map(),
+        }
+    }
+
+    /// Check if a file exists.
+    fn file_exists(&self, file_hash: &str) -> bool {
+        let record = self.files.get(&file_hash.to_string());
```

Comment:
We should not have to convert to string and then take a reference. We can take `&String` as a parameter if we have to keep strings.